sqlglot/optimizer/annotate_types.py

```diff
@@ -329,3 +339,20 @@ def _annotate_literal(self, expression):
     def _annotate_with_type(self, expression, target_type):
         expression.type = target_type
         return self._annotate_args(expression)
+
+    def _annotate_by_args(self, expression, *args):
+        self._annotate_args(expression)
+        expressions = []
+        for arg in args:
+            arg_expr = expression.args.get(arg)
+            if isinstance(arg_expr, list):
+                expressions += (expr for expr in arg_expr if expr)
+            elif arg_expr:
+                expressions.append(arg_expr)
```

Comment:
What about the following? In any case, I think we should prefer `extend` over `+=` here, because that's more consistent with sqlglot's codebase.

```suggestion
        for arg in args:
            arg_expr = expression.args.get(arg)
            expressions.extend(expr for expr in ensure_list(arg_expr) if expr)
```

Reply:
Seems cleaner indeed